### ISSUE TITLE
Duplicate messages using ConsumerGroup with kafkaHost

### ISSUE THREAD

**Issue by SimenB** [SOHU-Co/kafka-node#743]
## Bug Report

When trying to connect directly to the kafka brokers, all messages are duplicated. If I connect to zookeeper (using `host` instead of `kafkaHost`, no other changes to config (except for the value of this field, of course)) I only get a single message, as expected.

### Environment

* Node version: 6.11.2
* Kafka-node version: 2.2.0
* Kafka version: 0.10

### Include Sample Code to reproduce behavior

```js
const kafka = require('kafka-node');
const consumer = new kafka.ConsumerGroup({ kafkaHost: 'some host' }, 'test');
const zooConsumer = new kafka.ConsumerGroup({ host: 'some host' }, 'test');
const producer = new kafka.HighLevelProducer(new kafka.KafkaClient({ kafkaHost: 'some host' }));

consumer.on('message', msg => console.log(`${msg.value}`));
zooConsumer.on('message', msg => console.log(`from zoo ${msg.value}`));

producer.send([{ topic: 'test', messages: [{ type: 'type', values: { foo: 'bar' } }] }]);
```

`zooConsumer` is invoked twice, whilst `consumer` is only invoked once.

### Include output with [Debug](https://github.com/SOHU-Co/kafka-node#how-do-i-debug-an-issue) turned on

`kafkaHost`
```
  kafka-node:KafkaClient Connect attempt 1 +0ms
  kafka-node:KafkaClient Trying to connect to host: somehostport: 7794 +2ms
  kafka-node:KafkaClient Sending versions request to somehost:7794 +57ms
  kafka-node:KafkaClient broker socket connected {"host":"somehost","port":"7794"} +3ms
  kafka-node:KafkaClient Received versions response from somehost:7794 +30ms
  kafka-node:KafkaClient setting api support to {"produce":{"min":0,"max":2,"usable":2},"fetch":{"min":0,"max":3,"usable":0},"offset":{"min":0,"max":1,"usable":0},"metadata":{"min":0,"max":2,"usable":0},"leader":{"min":0,"max":0,"usable":false},"stopReplica":{"min":0,"max":0,"usable":false},"updateMetadata":{"min":0,"max":3,"usable":false},"controlledShutdown":{"min":1,"max":1,"usable":false},"offsetCommit":{"min":0,"max":2,"usable":2},"offsetFetch":{"min":0,"max":2,"usable":1},"groupCoordinator":{"min":0,"max":0,"usable":0},"joinGroup":{"min":0,"max":1,"usable":0},"heartbeat":{"min":0,"max":0,"usable":0},"leaveGroup":{"min":0,"max":0,"usable":0},"syncGroup":{"min":0,"max":0,"usable":0},"describeGroups":{"min":0,"max":0,"usable":false},"listGroups":{"min":0,"max":0,"usable":false},"saslHandshake":{"min":0,"max":0,"usable":false},"apiVersions":{"min":0,"max":0,"usable":0},"createTopics":{"min":0,"max":1,"usable":false},"deleteTopics":{"min":0,"max":0,"usable":false}} +1ms
  kafka-node:KafkaClient updating metadatas +874ms
  kafka-node:ConsumerGroup Connecting kafka-node-client +1ms
  kafka-node:ConsumerGroup GroupCoordinator Response: { coordinatorHost: 'somehost',
  coordinatorPort: 7794,
  coordinatorId: 2 } +64ms
  kafka-node:KafkaClient Sending versions request to somehost:7794 +54ms
  kafka-node:ConsumerGroup joinGroupResponse {"members":[{"subscription":["test"],"version":0,"id":"kafka-node-client-b68dd224-99b3-40ca-b41e-5c11e25d96f8"}],"generationId":43,"groupProtocol":"roundrobin","leaderId":"kafka-node-client-b68dd224-99b3-40ca-b41e-5c11e25d96f8","memberId":"kafka-node-client-b68dd224-99b3-40ca-b41e-5c11e25d96f8"} from kafka-node-client +27ms
  kafka-node:ConsumerGroup Assigning Partitions to members [ { subscription: [ 'test' ],
    version: 0,
    userData: undefined,
    id: 'kafka-node-client-b68dd224-99b3-40ca-b41e-5c11e25d96f8' } ] +1ms
  kafka-node:ConsumerGroup Using group protocol roundrobin +1ms
  kafka-node:ConsumerGroup loadingMetadata for topics: [ 'test' ] +2ms
  kafka-node:KafkaClient Received versions response from somehost:7794 +21ms
  kafka-node:KafkaClient setting api support to {"produce":{"min":0,"max":2,"usable":2},"fetch":{"min":0,"max":3,"usable":0},"offset":{"min":0,"max":1,"usable":0},"metadata":{"min":0,"max":2,"usable":0},"leader":{"min":0,"max":0,"usable":false},"stopReplica":{"min":0,"max":0,"usable":false},"updateMetadata":{"min":0,"max":3,"usable":false},"controlledShutdown":{"min":1,"max":1,"usable":false},"offsetCommit":{"min":0,"max":2,"usable":2},"offsetFetch":{"min":0,"max":2,"usable":1},"groupCoordinator":{"min":0,"max":0,"usable":0},"joinGroup":{"min":0,"max":1,"usable":0},"heartbeat":{"min":0,"max":0,"usable":0},"leaveGroup":{"min":0,"max":0,"usable":0},"syncGroup":{"min":0,"max":0,"usable":0},"describeGroups":{"min":0,"max":0,"usable":false},"listGroups":{"min":0,"max":0,"usable":false},"saslHandshake":{"min":0,"max":0,"usable":false},"apiVersions":{"min":0,"max":0,"usable":0},"createTopics":{"min":0,"max":1,"usable":false},"deleteTopics":{"min":0,"max":0,"usable":false}} +0ms
  kafka-node:ConsumerGroup mapTopicToPartitions { test: [ '0', '1', '2', '3', '4' ] } +5ms
  kafka-node:Roundrobin topicPartition: {"test":["0","1","2","3","4"]} +1ms
  kafka-node:Roundrobin groupMembers: [{"subscription":["test"],"version":0,"id":"kafka-node-client-b68dd224-99b3-40ca-b41e-5c11e25d96f8"}] +0ms
  kafka-node:Roundrobin members [ 'kafka-node-client-b68dd224-99b3-40ca-b41e-5c11e25d96f8' ] +1ms
  kafka-node:Roundrobin subscribers { 'kafka-node-client-b68dd224-99b3-40ca-b41e-5c11e25d96f8': [ 'test' ] } +0ms
  kafka-node:Roundrobin round robin on topic partition pairs:  [ { topic: 'test', partition: '0' },
  { topic: 'test', partition: '1' },
  { topic: 'test', partition: '2' },
  { topic: 'test', partition: '3' },
  { topic: 'test', partition: '4' } ] +1ms
  kafka-node:ConsumerGroup SyncGroup Request from kafka-node-client-b68dd224-99b3-40ca-b41e-5c11e25d96f8 +1ms
  kafka-node:ConsumerGroup SyncGroup Response +88ms
  kafka-node:ConsumerGroup kafka-node-client owns topics:  { test: [ 0, 1, 2, 3, 4 ] } +0ms
  kafka-node:ConsumerGroup kafka-node-client fetchOffset Response: {"test":{"0":79,"1":0,"2":0,"3":0,"4":0}} +34ms
  kafka-node:ConsumerGroup Has saved offsets +0ms
  kafka-node:ConsumerGroup generationId 43 +1ms
  kafka-node:ConsumerGroup kafka-node-client started heartbeats at every 10000 ms +3ms
  kafka-node:KafkaClient Sending versions request to somehost:7794 +25ms
  kafka-node:KafkaClient Sending versions request to somehost:7794 +1ms
  kafka-node:KafkaClient Sending versions request to somehost:7794 +22ms
  kafka-node:KafkaClient Received versions response from somehost:7794 +109ms
  kafka-node:KafkaClient setting api support to {"produce":{"min":0,"max":2,"usable":2},"fetch":{"min":0,"max":3,"usable":0},"offset":{"min":0,"max":1,"usable":0},"metadata":{"min":0,"max":2,"usable":0},"leader":{"min":0,"max":0,"usable":false},"stopReplica":{"min":0,"max":0,"usable":false},"updateMetadata":{"min":0,"max":3,"usable":false},"controlledShutdown":{"min":1,"max":1,"usable":false},"offsetCommit":{"min":0,"max":2,"usable":2},"offsetFetch":{"min":0,"max":2,"usable":1},"groupCoordinator":{"min":0,"max":0,"usable":0},"joinGroup":{"min":0,"max":1,"usable":0},"heartbeat":{"min":0,"max":0,"usable":0},"leaveGroup":{"min":0,"max":0,"usable":0},"syncGroup":{"min":0,"max":0,"usable":0},"describeGroups":{"min":0,"max":0,"usable":false},"listGroups":{"min":0,"max":0,"usable":false},"saslHandshake":{"min":0,"max":0,"usable":false},"apiVersions":{"min":0,"max":0,"usable":0},"createTopics":{"min":0,"max":1,"usable":false},"deleteTopics":{"min":0,"max":0,"usable":false}} +0ms
  kafka-node:KafkaClient Received versions response from somehost:7794 +2ms
  kafka-node:KafkaClient setting api support to {"produce":{"min":0,"max":2,"usable":2},"fetch":{"min":0,"max":3,"usable":0},"offset":{"min":0,"max":1,"usable":0},"metadata":{"min":0,"max":2,"usable":0},"leader":{"min":0,"max":0,"usable":false},"stopReplica":{"min":0,"max":0,"usable":false},"updateMetadata":{"min":0,"max":3,"usable":false},"controlledShutdown":{"min":1,"max":1,"usable":false},"offsetCommit":{"min":0,"max":2,"usable":2},"offsetFetch":{"min":0,"max":2,"usable":1},"groupCoordinator":{"min":0,"max":0,"usable":0},"joinGroup":{"min":0,"max":1,"usable":0},"heartbeat":{"min":0,"max":0,"usable":0},"leaveGroup":{"min":0,"max":0,"usable":0},"syncGroup":{"min":0,"max":0,"usable":0},"describeGroups":{"min":0,"max":0,"usable":false},"listGroups":{"min":0,"max":0,"usable":false},"saslHandshake":{"min":0,"max":0,"usable":false},"apiVersions":{"min":0,"max":0,"usable":0},"createTopics":{"min":0,"max":1,"usable":false},"deleteTopics":{"min":0,"max":0,"usable":false}} +0ms
  kafka-node:KafkaClient Received versions response from somehost:7794 +19ms
  kafka-node:KafkaClient setting api support to {"produce":{"min":0,"max":2,"usable":2},"fetch":{"min":0,"max":3,"usable":0},"offset":{"min":0,"max":1,"usable":0},"metadata":{"min":0,"max":2,"usable":0},"leader":{"min":0,"max":0,"usable":false},"stopReplica":{"min":0,"max":0,"usable":false},"updateMetadata":{"min":0,"max":3,"usable":false},"controlledShutdown":{"min":1,"max":1,"usable":false},"offsetCommit":{"min":0,"max":2,"usable":2},"offsetFetch":{"min":0,"max":2,"usable":1},"groupCoordinator":{"min":0,"max":0,"usable":0},"joinGroup":{"min":0,"max":1,"usable":0},"heartbeat":{"min":0,"max":0,"usable":0},"leaveGroup":{"min":0,"max":0,"usable":0},"syncGroup":{"min":0,"max":0,"usable":0},"describeGroups":{"min":0,"max":0,"usable":false},"listGroups":{"min":0,"max":0,"usable":false},"saslHandshake":{"min":0,"max":0,"usable":false},"apiVersions":{"min":0,"max":0,"usable":0},"createTopics":{"min":0,"max":1,"usable":false},"deleteTopics":{"min":0,"max":0,"usable":false}} +0ms
```

`host`:
```
  kafka-node:ConsumerGroup Connecting kafka-node-client +0ms
  kafka-node:ConsumerGroup GroupCoordinator Response: { coordinatorHost: 'somehost',
  coordinatorPort: 7794,
  coordinatorId: 2 } +59ms
  kafka-node:ConsumerGroup joinGroupResponse {"members":[{"subscription":["test"],"version":0,"id":"kafka-node-client-9ac3480b-f72c-4412-970d-ce52b1119507"}],"generationId":45,"groupProtocol":"roundrobin","leaderId":"kafka-node-client-9ac3480b-f72c-4412-970d-ce52b1119507","memberId":"kafka-node-client-9ac3480b-f72c-4412-970d-ce52b1119507"} from kafka-node-client +98ms
  kafka-node:ConsumerGroup Assigning Partitions to members [ { subscription: [ 'test' ],
    version: 0,
    userData: undefined,
    id: 'kafka-node-client-9ac3480b-f72c-4412-970d-ce52b1119507' } ] +1ms
  kafka-node:ConsumerGroup Using group protocol roundrobin +1ms
  kafka-node:ConsumerGroup loadingMetadata for topics: [ 'test' ] +2ms
  kafka-node:ConsumerGroup mapTopicToPartitions { test: [ '0', '1', '2', '3', '4' ] } +33ms
  kafka-node:Roundrobin topicPartition: {"test":["0","1","2","3","4"]} +0ms
  kafka-node:Roundrobin groupMembers: [{"subscription":["test"],"version":0,"id":"kafka-node-client-9ac3480b-f72c-4412-970d-ce52b1119507"}] +0ms
  kafka-node:Roundrobin members [ 'kafka-node-client-9ac3480b-f72c-4412-970d-ce52b1119507' ] +1ms
  kafka-node:Roundrobin subscribers { 'kafka-node-client-9ac3480b-f72c-4412-970d-ce52b1119507': [ 'test' ] } +0ms
  kafka-node:Roundrobin round robin on topic partition pairs:  [ { topic: 'test', partition: '0' },
  { topic: 'test', partition: '1' },
  { topic: 'test', partition: '2' },
  { topic: 'test', partition: '3' },
  { topic: 'test', partition: '4' } ] +0ms
  kafka-node:ConsumerGroup SyncGroup Request from kafka-node-client-9ac3480b-f72c-4412-970d-ce52b1119507 +1ms
  kafka-node:ConsumerGroup SyncGroup Response +93ms
  kafka-node:ConsumerGroup kafka-node-client owns topics:  { test: [ 0, 1, 2, 3, 4 ] } +0ms
  kafka-node:ConsumerGroup kafka-node-client fetchOffset Response: {"test":{"0":81,"1":0,"2":0,"3":0,"4":0}} +29ms
  kafka-node:ConsumerGroup Has saved offsets +0ms
  kafka-node:ConsumerGroup generationId 45 +0ms
  kafka-node:ConsumerGroup kafka-node-client started heartbeats at every 10000 ms +2ms
```

**Comment by hyperlink** [comment]
The  producer send need a callback addded and the consumerGroups should use `fromOffset: 'earliest'` since the message is likely sent before the consumers are ready. After these changes I only get one message since they are from the same group.

```js
const kafka = require('kafka-node');
const consumerOptions = {
  fromOffset: 'earliest'
};
const consumer = new kafka.ConsumerGroup(Object.assign({ kafkaHost: '127.0.0.1:9092' }, consumerOptions), 'test');
const zooConsumer = new kafka.ConsumerGroup(Object.assign({ host: '127.0.0.1:2181' }, consumerOptions), 'test');
const producer = new kafka.HighLevelProducer(new kafka.KafkaClient({ kafkaHost: '127.0.0.1:9092' }));

consumer.on('message', msg => console.log('konsumer', msg));
zooConsumer.on('message', msg => console.log('zooumer', msg));

console.log('sending message');
producer.send([{ topic: 'test', messages: [{ type: 'type', values: { foo: 'bar' } }] }], function (error) {
  if (error) {
    console.error('producer error', error);
  }
});
```

**Comment by SimenB** [comment]
I'll test tomorrow. We also got 2 messages when producing using the java client, but I can try tweaking the offset :smile:
Thanks!

**Comment by SimenB** [comment]
Do I need the callback to send? Why doesn't that emit error instead?

**Comment by SimenB** [comment]
I get a single message if both of the consumers are used at the same time. If I comment out the ZK one, I get 2 message, if I comment out the kafka one, I get 1 message.

Changing `offset` makes no difference, and it happens if we produce the message from a java client (deleting the producer in this example)

**Comment by hyperlink** [comment]
@SimenB I can't reproduce this. Are you able to reproduce it using the docker setup we have in the project?

`KAFKA_VERSION=0.10 ./start-docker.sh`

**Comment by SimenB** [comment]
Seems to be an error with our cluster. If I fire up the docker stuff from this repo, I get a single message. Just switching `kafkaHost` to ours and keeping all other code results in 2 messages. Thanks for the help!

**Comment by SimenB** [comment]
@hyperlink I was able to reproduce the error by having a topic with replication factor of 2. I'm having issues with getting 2 brokers running with docker compose, but are you able to reproduce if the `test` topic has a replication factor of 2 (or higher)?

**Comment by hyperlink** [comment]
I'll look into setting up two brokers to test this.

**Comment by hyperlink** [comment]
Still unable to reproduce using replication factor of 2.

Here's my `docker-compose` setup. You will need to run your test in a separate container on same docker network.

```yml
version: '2'
services:
  zookeeper:
    image: jplock/zookeeper
    ports:
      - "2181:2181"
  kafka1:
    image: wurstmeister/kafka:0.10.2.1
    ports:
      - "9092"
    depends_on:
      - zookeeper
    environment:
      KAFKA_ADVERTISED_HOST_NAME: kafka1
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_LISTENERS: "PLAINTEXT://kafka1:9092"
      KAFKA_ADVERTISED_LISTENERS: "PLAINTEXT://kafka1:9092"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
  kafka2:
    image: wurstmeister/kafka:0.10.2.1
    ports:
      - "9092"
    depends_on:
      - zookeeper
    environment:
      KAFKA_ADVERTISED_HOST_NAME: kafka2
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_LISTENERS: "PLAINTEXT://kafka2:9092"
      KAFKA_ADVERTISED_LISTENERS: "PLAINTEXT://kafka2:9092"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
  kafka3:
    image: wurstmeister/kafka:0.10.2.1
    ports:
      - "9092"
    depends_on:
      - zookeeper
    environment:
      KAFKA_ADVERTISED_HOST_NAME: kafka3
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_LISTENERS: "PLAINTEXT://kafka3:9092"
      KAFKA_ADVERTISED_LISTENERS: "PLAINTEXT://kafka3:9092"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
```

Manually create the topic.

```
bash-4.3# ./kafka-topics.sh --create --zookeeper zookeeper:2181 --topic e936c277-e1ec-4988-a662-e57a67263c91 --partitions 2 --replication-factor 2
Created topic "e936c277-e1ec-4988-a662-e57a67263c91".
bash-4.3# ./kafka-topics.sh --describe --zookeeper zookeeper:2181 --topic e936c277-e1ec-4988-a662-e57a67263c91
Topic:e936c277-e1ec-4988-a662-e57a67263c91	PartitionCount:2	ReplicationFactor:2	Configs:
	Topic: e936c277-e1ec-4988-a662-e57a67263c91	Partition: 0	Leader: 1002	Replicas: 1002,1003	Isr: 1002,1003
	Topic: e936c277-e1ec-4988-a662-e57a67263c91	Partition: 1	Leader: 1003	Replicas: 1003,1001	Isr: 1003,1001
```

then ran your test case.

```js
const kafka = require('.');
const consumerOptions = {
  fromOffset: 'earliest',
  groupId: '1234'
};

const topic = 'e936c277-e1ec-4988-a662-e57a67263c91';

const consumer = new kafka.ConsumerGroup(Object.assign({ kafkaHost: 'kafka1:9092' }, consumerOptions), topic);
// const zooConsumer = new kafka.ConsumerGroup(Object.assign({ host: 'zookeeper:2181' }, consumerOptions), 'test');
const producer = new kafka.HighLevelProducer(new kafka.KafkaClient({ kafkaHost: 'kafka1:9092' }));

consumer.on('message', msg => console.log('konsumer', msg));
// zooConsumer.on('message', msg => console.log('zooumer', msg));

console.log('sending message');
producer.send([{ topic: topic, messages: [{ type: 'type', values: { foo: 'bar' } }] }], function (error) {
  if (error) {
    console.error('producer error', error);
  }
});
```


```
docker run -it --rm --network docker_default --name node_test -v "$PWD":/usr/src/app -w /usr/src/app node:6 node issue-743.js
```

```
sending message
konsumer { topic: 'e936c277-e1ec-4988-a662-e57a67263c91',
  value: '[object Object]',
  offset: 0,
  partition: 0,
  highWaterOffset: 1,
  key: null }
```

**Comment by SimenB** [comment]
Not the easiest this, but this should do it!

Same docker compose as you.

```sh-session
$ docker-compose up -d
# ...wait
$ docker exec -it kafkanode_kafka1_1 /opt/kafka/bin/kafka-topics.sh --create --zookeeper zookeeper:2181 --replication-factor 2 --partitions 5 --topic test
```

`consumer.js`
```js
'use strict';

const kafka = require('./');
const consumer = new kafka.ConsumerGroup({ kafkaHost: 'kafka1:9092'}, 'test');

consumer.on('message', msg => console.log('konsumer', msg));
```

`producer.js`
```js
'use strict';

const kafka = require('./');

const producer = new kafka.HighLevelProducer(new kafka.KafkaClient({ kafkaHost: 'kafka1:9092' }));

producer.on('ready', () => {
  console.log('sending message');

  producer.send([{ topic: 'test', messages: [{ type: 'type', values: { foo: 'bar' } }] }], error => {
    if (error) {
      console.error('producer error', error);
    }

    process.exit(0);
  });
});
```

```sh-session
$ docker run -it --rm --network kafkanode_default --name node_test2 -v "$PWD":/usr/src/app -w /usr/src/app node:6 node consumer.js
```

Other terminal:
```sh-session
$ docker run --rm --network kafkanode_default -v "$PWD":/usr/src/app -w /usr/src/app node:6 node producer.js
```

Run the producer code twice (allow the first message to be printed before sending a new one).

The _first_ message on a new topic only shows up once, every single one after that (even restarting the consumer image) is duplicated.

**Comment by hyperlink** [comment]
Looks like this is related to support for different versions of the protocol. I will look into fixing it for now you can workaround this by disabling versions in the `KafkaClient`.

Add this options to your consumer group options.

```
  versions: {
    disabled: true
  }
```

**Comment by SimenB** [comment]
Great that you could reproduce it.

Thanks for the workaround, I'll test it Monday 🙂

**Comment by SimenB** [comment]
Cool! Can you do a patch release?

**Comment by hyperlink** [comment]
You verified master branch works for you?

**Comment by SimenB** [comment]
I've verified that setting `versions: { disabled: true }` works. Can verify master now

**Comment by SimenB** [comment]
`master` works correctly 🙂

**Comment by hyperlink** [comment]
Published as 2.2.1. Thanks for your help @SimenB.
